app/queries/internal/moderators_query.rb

```diff
@@ -4,34 +4,21 @@ class ModeratorsQuery
       state: :trusted
     }.with_indifferent_access.freeze
 
-    def self.call(relation: User.all, options: {})
+    def self.call(options: {})
       options = DEFAULT_OPTIONS.merge(options)
       state, search = options.values_at(:state, :search)
 
-      relation = if state.to_s == "potential"
-                   relation.where(
-                     "id NOT IN (SELECT user_id FROM users_roles WHERE role_id = ?)",
-                     role_id_for(:trusted),
-                   ).order("users.comments_count DESC")
-                 else
-                   relation.joins(:users_roles).
-                     where(users_roles: { role_id: role_id_for(state) })
-                 end
+      formatted_params = if state.to_s == "potential"
+                           { exclude_roles: ["trusted"], sort_by: "comments_count", sort_direction: "desc" }
+                         else
+                           { roles: [state] }
+                         end
 
-      relation = search_relation(relation, search) if search.presence
+      formatted_params = formatted_params.merge(name_fields: search) if search.presence
+      results = Search::User.search_documents(params: formatted_params.merge(options.slice(:page, :per_page)))
+      user_ids = results.map { |doc| doc["id"] }
 
-      relation
-    end
-
-    def self.role_id_for(role)
-      Role.find_by!(name: role).id
-    end
-
-    def self.search_relation(relation, search)
-      relation.where(
-        "users.username ILIKE :search OR users.name ILIKE :search",
-        search: "%#{search}%",
-      )
+      Kaminari.paginate_array(User.find(user_ids), total_count: results.first&.dig("total"))
```

Comment:
You can skip this part and pass the pagination params to our Search class and it will paginate for you

Reply:
Yeah, I'm doing that at line 18.

The issue is that some internal pages use `Kaminari` pagination helper that requires some specific attributes which an array doesn't have.

I rather use the doc results from the `elasticsearch` and not use AR to search again for the records. But for that, I might need to change the `search_documents` results structure for something like:

```ruby
{
  total_count: 100,
  page: 1,
  per_page: 20,
  pages: 5,
  users: [
    { id: 1, name: 'User Name', roles: [] },
    { id: 2, name: 'User Name 2', roles: []}
  ]
}
```

That way we could easily create pagination on front-end using direct `elasticsearch` results and also instantiate User objects from its attributes or just use OpenStruct if we just need to access the values.

```ruby
results = Search::User.search_documents(params: {})
users = results['users'].map |attr|
  OpenStruct.new(attr)
end
```

Something around those lines. WDYT?

Reply:
I like the idea of returning additional metadata with the search documents. I could see that being useful in many places.